File: SDE_sheet/Arrays/17_unique_paths/solution.ts
function uniquePaths(m: number, n: number): number {
    const dp: number[][] = new Array(m);
    
    for (let i = 0; i < m; i++) {
        dp[i] = new Array(n);
    }
    
    for (let i: number = 0; i < m; i++) {
        for (let j: number = 0; j < n; j++) {
            if (i === 0 || j === 0) {
                dp[i][j] = 1;
            } else {
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }
    }
    
    return dp[m - 1][n - 1];
}

File: SDE_sheet/Arrays/17_unique_paths/solution.rs
fn unique_paths(m: usize, n: usize) -> i32 {
    let mut dp = vec![vec![0; n]; m];
    
    for i in 0..m {
        for j in 0..n {
            if i == 0 || j == 0 {
                dp[i][j] = 1;
            } else {
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }
    }
    
    dp[m - 1][n - 1]
}

